hazelcast/src/main/java/com/hazelcast/query/PartitionPredicate.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+/**
+ * A {@link Predicate} that restricts the execution of a Predicate to a single Partition. This can help to speed up query
+ * execution since only 1 instead of all partitions needs to be queried.
+ *
+ * This predicate can only be used as an outer predicate.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class PartitionPredicate<K, V> implements Predicate<K, V>, DataSerializable {
+
```

Comment:
This is mixing concerns. I understand it was easier to implement it this way that you just wrap a predicate around, since the existing implementation didn't have to be changed at all, but why should even a user know about this class at all? This is just an implementation detail.
It you want to expose it like this, the PagingPredicate  should be a private class hidden behind the Predicates factory, exposed through the Predicate Interface only.
Like this:
```
Predicate query = Predicates.equals("age", 40)
Predicate query = Predicates.onPartition(query, partitionKey)
map.values(query)
```

To me, however, adding a logic on how many partitions a query should be executed should not be passed by a predicate. I know it's just a wrapper, but it's not an healthy way how to grow an API in a sustainable way. What if I want to add one or two similar flags, how many times will I wrap the predicate around?

Don't get me wrong, I like the idea, but not how it's exposed to the end-user in the public API.
Predicates may be used in multiple places. e.g. ReplicatedMap, TransactionalMap, etc. 
What will happen if the user passes a PartitionPredicate there?
Even in the IMap you changed `keySet`, `entrySet` and `values` only. 
There are other methods like
`localKeySet` , `executeOnEntries`  that do not support this type of Predicate.